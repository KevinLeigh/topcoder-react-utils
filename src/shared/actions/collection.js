/**
 * Proof-of-concept code. Will be documented once stable. In case of any
 * questions before that, feel free to contact `birdofpreyru`.
 */

import _ from 'lodash';
import * as redux from 'utils/redux';

/**
 * Adds items to the collection.
 * @param {Object} index Item ID > item mapping for the items to add.
 * @return {Object}
 */
function addItems(index) {
  return { index, timestamp: Date.now() };
}

/**
 * Increments counters of item references.
 * @param {String|String[]} ids
 * @return {String[]}
 */
function bookItems(ids) {
  return _.isArray(ids) ? ids : [ids];
}

/**
 * Removes from the state all unuzed items with timestamps older than maxage.
 * @param {Number} timestamp
 */
function clean(timestamp) {
  return timestamp;
}

/**
 * Decreases counters of item references.
 * @param {String|String[]} ids
 * @return {String[]}
 */
function freeItems(ids) {
  return _.isArray(ids) ? ids : [ids];
}

/**
 * Inits loading of an item.
 * @param {String} operationId
 * @param {String} itemId
 * @return {Object}
 */
function loadItemInit(operationId, itemId) {
  return { itemId, operationId };
}

/**
 * Finalizes loading of an item.
 * @param {String} operationId
 * @param {String} itemId
 * @param {Any} item
 * @return {Object}
 */
function loadItemDone(operationId, itemId, item) {
  return {
    item,
    itemId,
    operationId,
    timestamp: Date.now(),
  };
}

export default redux.createActions({
  COLLECTION: {
    ADD_ITEMS: addItems,
    BOOK_ITEMS: bookItems,
    CLEAN: clean,
    FREE_ITEMS: freeItems,
    LOAD_ITEM_INIT: loadItemInit,
    LOAD_ITEM_DONE: loadItemDone,
  },
});
