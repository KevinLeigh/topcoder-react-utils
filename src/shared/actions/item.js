/**
 * @module "actions.item"
 * @desc Actions for management of item data in Redux store.
 */

import * as redux from 'utils/redux';

/**
 * @static
 * @desc Creates an action that drops item's data if (i) the value of item's
 *  reference counter is not positive; (ii) item's timestamp is older than
 *  `olderThan` value.
 * @param {Number} olderThan=Number.MAX_VALUE Optional. Timestamp [ms].
 *  If provided, item's  data will be dropped only if they were received / set
 *  prior to this moment in time.
 * @return {Action}
 */
function dropData(olderThan = Number.MAX_VALUE) {
  return olderThan;
}

/**
 * @static
 * @desc Creates an action that signals the beginning of async loading of
 *  item's data.
 * @param {Number|String} loadingOperationId The unique identifier of the
 *  loading operation. It will be written into item's segment of Redux store
 *  with the following efffects:
 *
 *  - We consider that item's data are being loaded if, and only if,
 *    the item's `loadingOperationId` has a truly value;
 *
 *  - `loadingOperationId` is used to match the corresponding
 *    {@link module:actions.item.loadDataDone} operation.
 *
 * @return {Action}
 */
function loadDataInit(loadingOperationId) {
  return loadingOperationId;
}

/**
 * @static
 * @desc Creates an action that handles the result of async loading of item's
 *  data.
 * @param {Number|String} loadingOperationId The unique identifier of the
 *  loading operation. If it does not match the value currently written for
 *  the item in the Redux store (see {@link module:actions.item.loadDataInit}),
 *  the action will be silently ignored.
 * @param {Any} data Any value to store as the item's data.
 * @return {Action}
 */
function loadDataDone(loadingOperationId, data) {
  return { data, loadingOperationId, timestamp: Date.now() };
}

/**
 * @static
 * @desc Creates an action that stores given data into the item.
 * @param {Any} data Any value to store as the item's data.
 * @return {Action}
 */
function setData(data) {
  return { data, timestamp: Date.now() };
}

/**
 * @static
 * @desc Creates an action that updates the item's reference counter.
 * @param {Number} shift The value to add to the item's reference counter.
 *  Should be `1` or `-1` in majority of cases, although any other values
 *  are permitted.
 * @return {Action}
 */
function updateReferenceCounter(shift) {
  return shift;
}

export default redux.createActions({
  ITEM: {
    DROP_DATA: dropData,
    LOAD_DATA_INIT: loadDataInit,
    LOAD_DATA_DONE: loadDataDone,
    SET_DATA: setData,
    UPDATE_REFERENCE_COUNTER: updateReferenceCounter,
  },
});
