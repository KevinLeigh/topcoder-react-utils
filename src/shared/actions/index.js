import collection from './collection';
import item from './item';

export default {
  ...collection,
  ...item,
};
