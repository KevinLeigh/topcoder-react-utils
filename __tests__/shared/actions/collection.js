/**
 * Testing of the collection actions.
 */

import actions from 'actions/collection';
import MockDate from 'mockdate';

const a = actions.collection;

beforeAll(() => MockDate.set('2018-05-25T19:03:35+00:00'));

afterAll(() => MockDate.reset());

test('Exports expected actions', () => expect(actions).toMatchSnapshot());

test('`addItems` payload creator', () =>
  expect(a.addItems({ id1: 'Item 1', id2: 'Item 2' })).toMatchSnapshot());

test('`bookItems` payload creator', () => {
  expect(a.bookItems('id1')).toMatchSnapshot();
  expect(a.bookItems('id1', 7)).toMatchSnapshot();
  expect(a.bookItems(['id1', 'id2', 'id3'])).toMatchSnapshot();
});

test('`clean` payload creator', () => {
  expect(a.clean()).toMatchSnapshot();
  expect(a.clean(100)).toMatchSnapshot();
});

test('`freeItems` payload creator', () => {
  expect(a.freeItems('id1')).toMatchSnapshot();
  expect(a.freeItems('id1', 7)).toMatchSnapshot();
  expect(a.freeItems(['id1', 'id2', 'id3'])).toMatchSnapshot();
});

test('`loadItemInit` payload creator', () =>
  expect(a.loadItemInit('operation-id', 'item-id')).toMatchSnapshot());

test('`loadItemDone` payload creator', () =>
  expect(a.loadItemDone('operation-id', 'item-id', 'item')).toMatchSnapshot());
