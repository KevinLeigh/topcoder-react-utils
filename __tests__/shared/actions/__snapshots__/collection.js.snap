// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`addItems\` payload creator 1`] = `
Object {
  "payload": Object {
    "index": Object {
      "id1": "Item 1",
      "id2": "Item 2",
    },
    "timestamp": 1527275015000,
  },
  "type": "COLLECTION/ADD_ITEMS",
}
`;

exports[`\`bookItems\` payload creator 1`] = `
Object {
  "payload": Object {
    "factor": 1,
    "ids": Array [
      "id1",
    ],
  },
  "type": "COLLECTION/BOOK_ITEMS",
}
`;

exports[`\`bookItems\` payload creator 2`] = `
Object {
  "payload": Object {
    "factor": 7,
    "ids": Array [
      "id1",
    ],
  },
  "type": "COLLECTION/BOOK_ITEMS",
}
`;

exports[`\`bookItems\` payload creator 3`] = `
Object {
  "payload": Object {
    "factor": 1,
    "ids": Array [
      "id1",
      "id2",
      "id3",
    ],
  },
  "type": "COLLECTION/BOOK_ITEMS",
}
`;

exports[`\`clean\` payload creator 1`] = `
Object {
  "payload": 1.7976931348623157e+308,
  "type": "COLLECTION/CLEAN",
}
`;

exports[`\`clean\` payload creator 2`] = `
Object {
  "payload": 100,
  "type": "COLLECTION/CLEAN",
}
`;

exports[`\`freeItems\` payload creator 1`] = `
Object {
  "payload": Object {
    "factor": 1,
    "ids": Array [
      "id1",
    ],
  },
  "type": "COLLECTION/FREE_ITEMS",
}
`;

exports[`\`freeItems\` payload creator 2`] = `
Object {
  "payload": Object {
    "factor": 7,
    "ids": Array [
      "id1",
    ],
  },
  "type": "COLLECTION/FREE_ITEMS",
}
`;

exports[`\`freeItems\` payload creator 3`] = `
Object {
  "payload": Object {
    "factor": 1,
    "ids": Array [
      "id1",
      "id2",
      "id3",
    ],
  },
  "type": "COLLECTION/FREE_ITEMS",
}
`;

exports[`\`loadItemDone\` payload creator 1`] = `
Object {
  "payload": Object {
    "item": "item",
    "itemId": "item-id",
    "operationId": "operation-id",
    "timestamp": 1527275015000,
  },
  "type": "COLLECTION/LOAD_ITEM_DONE",
}
`;

exports[`\`loadItemInit\` payload creator 1`] = `
Object {
  "payload": Object {
    "itemId": "item-id",
    "operationId": "operation-id",
  },
  "type": "COLLECTION/LOAD_ITEM_INIT",
}
`;

exports[`Exports expected actions 1`] = `
Object {
  "collection": Object {
    "addItems": [Function],
    "bookItems": [Function],
    "clean": [Function],
    "freeItems": [Function],
    "loadItemDone": [Function],
    "loadItemInit": [Function],
  },
}
`;
